hack: allow pkg/ to import internal/ packages

corhere opened the pull request:
The pkg-imports validation prevents reusable library packages from depending on the whole daemon, accidentally or intentionally. The allowlist is overly restrictive as it also prevents us from reusing code in both `pkg/` and `daemon/` unless that code is also made into a reusable library package under `pkg/`. Allow `pkg/` packages to import `internal/` packages which do not transitively depend on disallowed packages.

thaJeztah commented on pkg/urlutil/deprecated.go:
```diff
@@ -8,14 +8,12 @@ import "github.com/docker/docker/builder/remotecontext/urlutil"
 // to detect build-context type, or use strings.HasPrefix() to check if the
 // string has a https:// or http:// prefix.
 func IsURL(str string) bool {
-	// TODO(thaJeztah) when removing this alias, remove the exception from hack/validate/pkg-imports and hack/make.ps1 (Validate-PkgImports)
```
We can probably merge https://github.com/moby/moby/pull/44261 first 😃

thaJeztah commented:
I think the reason wasn't only generated code, but also to prevent (e.g.) a `pkg/` importing `daemon/xx` code, which generally isn't desirable. Originally the `pkg/` directory was similar to how separate (sub)modules would now be; you want those to be (mostly) self-contained, and not depend on other (potentially _"internal"_ ) packages.

thaJeztah commented on hack/make.ps1:
```diff
@@ -256,20 +256,12 @@ Function Validate-PkgImports($headCommit, $upstreamCommit) {
-        $imports = @() + $imports -NotMatch "^github.com/docker/docker/pkg/" `
-                                  -NotMatch "^github.com/docker/docker/vendor" `
-                                  -Match "^github.com/docker/docker" `
+        $imports = @() + $imports -Match "^github.com/docker/docker/(cli|cmd)" `
```
I wonder if instead of our own crafted check, if we could replace it with `depguard` 🤔 (with some proper configuration of that) https://github.com/moby/moby/blob/058010187bdae5364d82c4cff204f103231b13e0/hack/validate/golangci-lint.yml#L29-L35

(we can probably remove that specific check though, as `io/ioutils` is marked deprecated, so should probably already show up in other linters)

corhere replied:
https://golangci-lint.run/usage/linters#depguard
Yeah, it does look feasible with something like
```yaml
linters-settings:
  depguard:
    additional-guards:
      - list-type: denylist
        packages:
          - github.com/docker/docker/cmd/**
          - github.com/docker/docker/cli/**
          - github.com/docker/docker/daemon/**
        ignore-file-rules:
          - "!pkg/**/*.go"
```

thaJeztah replied:
Ah, nice! Yes, if it works, I like that approach; it's more visible and easier to understand to others (without having to dig into obscure shell scripts)

corhere commented:
> you want those to be (mostly) self-contained, and not depend on other (potentially _"internal"_ ) packages.

My motivation is literally to allow `github.com/docker/docker/internal/...` packages to be importable from `pkg` so that the internal packages can also be imported by `daemon`. I'm of the opinion that the `pkg/` convention should really be about [communicating that these packages are explicitly safe for use by others](https://github.com/golang-standards/project-layout#pkg) and nothing more. There is nothing inherently unsafe about those packages depending on _"internal"_ packages as part of their implementation details.

That all being said, I am onboard with disallowing _application_ packages from being imported by `pkg/` packages. Depending on `cmd/...`, `cli/...` or `daemon/...` packages most likely means one of two things: either it's a misguided use of application-specific code which is not appropriate to use in library code, or it's a signal that the code really is reusable library code that should be moved into a package more befitting of library code.

corhere committed:
hack: allow pkg/ to import internal/ packages

The pkg-imports validation prevents reusable library packages from
depending on the whole daemon, accidentally or intentionally. The
allowlist is overly restrictive as it also prevents us from reusing code
in both pkg/ and daemon/ unless that code is also made into a reusable
library package under pkg/. Allow pkg/ packages to import internal/
packages which do not transitively depend on disallowed packages.

Signed-off-by: Cory Snider <csnider@mirantis.com>
